### PR TITLE
Iteration 5

### PR DESCRIPTION
**Pull request by jasnell**
Starting iteration-5...

rolled in changes from https://github.com/nodejs/quic/pull/26 and https://github.com/nodejs/quic/pull/25 so those don't need to be rebased. /cc @danbev @gengjiawen 

##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [ ] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

File: lib/internal/quic/core.js [hunk 1]
```diff
@@ -455,14 +460,14 @@ function connectAfterLookup(
...
     _createClientSession(
-        this, type,
-        ip,
-        port || 0,
-        flags,
-        clientSecureContext.context,
-        servername || ip,
-        remoteTransportParams,
-        sessionTicket);
+      this, type,
```
Comment (BridgeAR) [inline lib/internal/quic/core.js]: Nit: these two entries should be aligned on two lines instead of one.

**Comment by jasnell** [comment]
I'm stepping back and reworking this one...

**Comment by jasnell** [comment]
Ok, @danbev @mcollina @nodejs/quic ... please take a look. There's still tons to do

File: src/env-inl.h [hunk 2]
```diff
@@ -613,6 +613,12 @@ inline void Environment::set_quic_state(std::unique_ptr<QuicState> buffer) {
   quic_state_ = std::move(buffer);
 }
 
+inline quic::QuicMonitor* Environment::quic_monitor() {
+  if (!quic_monitor_)
+    quic_monitor_.reset(new quic::QuicMonitor(this));
```
Comment (gengjiawen) [inline src/env-inl.h]: Clang-tidy issue: 
`quic_monitor_ = std::make_unique<quic::QuicMonitor>(this);`

File: src/node_quic_session.cc [hunk 3]
```diff
@@ -3053,26 +2315,39 @@ void QuicClientSession::SetLocalCryptoLevel(ngtcp2_crypto_level level) {
...
   }
-  return new QuicClientSession(socket,
-                               obj,
-                               addr,
-                               version,
-                               context,
-                               hostname,
-                               port);
+
+  session.reset(
```
Comment (gengjiawen) [inline src/node_quic_session.cc]: clang-tidy issue: use make-shared
```cpp
  session = std::make_shared<QuicClientSession>(socket,
                                                obj,
                                                addr,
                                                version,
                                                context,
                                                hostname,
                                                port,
                                                early_transport_params,
                                                session_ticket);
```

**Comment by jasnell** [comment]
Planning to get this landed today so we can keep iterating forward.

**Comment by jasnell** [comment]
Landed

**Pull request closed without merging**